packages/smooth_app/lib/pages/input/brand_suggestion_manager.dart

```diff
@@ -0,0 +1,62 @@
+import 'dart:async';
+
+import 'package:openfoodfacts/openfoodfacts.dart';
+import 'package:smooth_app/query/product_query.dart';
+
+/// Manager that returns the elastic suggestions for the latest brand input.
+///
+/// See also: [SuggestionManager].
+class BrandSuggestionManager {
+  BrandSuggestionManager({
+    this.limit = 25,
+    this.user,
+  });
+
+  final int limit;
+  final User? user;
+
+  final List<String> _inputs = <String>[];
+  final Map<String, List<String>> _cache = <String, List<String>>{};
+
+  /// Returns suggestions about the latest input.
+  Future<List<String>> getSuggestions(
+    final String input,
+  ) async {
+    _inputs.add(input);
+    final List<String>? cached = _cache[input];
+    if (cached != null) {
+      return cached;
+    }
+    final AutocompleteSearchResult result =
+        await OpenFoodSearchAPIClient.autocomplete(
+      query: input,
+      taxonomyNames: <TaxonomyName>[TaxonomyName.brand],
+      // for brands, language must be English
+      language: OpenFoodFactsLanguage.ENGLISH,
+      // for the moment, only TEST is available but it's good enough
+      uriHelper: uriHelperFoodTest,
+      user: ProductQuery.getUser(),
+      size: limit,
+      fuzziness: Fuzziness.none,
+    );
+    final List<String> tmp = <String>[];
+    if (result.options != null) {
+      for (final AutocompleteSingleResult option in result.options!) {
+        final String text = option.text;
+        if (!tmp.contains(text)) {
+          tmp.add(text);
+        }
+      }
+    }
+    _cache[input] = tmp;
+    // meanwhile there might have been some calls to this method, adding inputs.
+    for (final String latestInput in _inputs.reversed) {
```

Comment:
By curiosity, why do you in the reverse order?
(I think it needs some comments to better understand why).

Reply:
@g123k Will be dealt with by https://github.com/openfoodfacts/openfoodfacts-dart/pull/851
For the record the "reverse" is the core of the class and explained in the class comments: regardless of the result - that may have been delayed - what we really want is what the user asked _lastly_, which is the first in the reverse order.